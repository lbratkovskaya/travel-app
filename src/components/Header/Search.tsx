import React from 'react';
import { useTranslation } from 'react-i18next';
import { makeStyles, TextField, InputAdornment } from '@material-ui/core';
import { useDispatch, useSelector } from 'react-redux';
import SearchIcon from '@material-ui/icons/Search';
import { IAppState } from '../../store/types';
import { findCountry } from '../../controller/utils';

const useStyles = makeStyles(() => ({
  button: {
    background: 'none',
    border: 'none',
    cursor: 'pointer',
    '&:hover': {
      outline: 'none',
      border: 'none',
      boxShadow: 'none',
      opacity: '0.5',
    },
    '&:focus': {
      outline: 'none',
      border: 'none',
      boxShadow: 'none',
    },
    '&:active': {
      outline: 'none',
      border: 'none',
      boxShadow: 'none',
    },
  },
  input: {
    backgroundColor: 'white',
  },
}));

const Search: React.FC = () => {
  const classes = useStyles();
  const dispatch = useDispatch();
  const countries = useSelector((state: IAppState) => state.countriesList);
  const lang = useSelector((state: IAppState) => state.lang);
  const { t } = useTranslation();

  return (
    <TextField
      className={classes.input}
      autoFocus
      placeholder={t('search')}
      type="search"
      variant="outlined"
      margin="normal"
      InputProps={{
        startAdornment: (
          <InputAdornment position="start">
            <button className={classes.button} type="button">
              <SearchIcon />
            </button>
          </InputAdornment>
        ),
      }}
      onChange={(event) => {
        dispatch(
          {
            type: 'FILTER_COUNTRIES',
            payload: { countries: findCountry(event.currentTarget.value, countries, lang) },
          },
        );
      }}
    />
  );
};

export default Search;
