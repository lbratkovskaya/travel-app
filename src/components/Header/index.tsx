import React from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import {
  fade,
  makeStyles,
  AppBar,
  Toolbar,
  IconButton,
  Typography,
  MenuItem,
  Menu,
  Avatar,
} from '@material-ui/core';
import MoreIcon from '@material-ui/icons/MoreVert';
import { useParams } from 'react-router-dom';
import handleLangChange from '../../controller/handlers';
import { IAppState } from '../../store/types';
import SignInForm from '../RegForms/SignInForm';
import SignUpForm from '../RegForms/SignUpForm';
import Search from './Search';
import { URLParamTypes } from '../../types';

const useStyles = makeStyles((theme) => ({
  appBar: {
    background: 'none',
    boxShadow: 'none',
    display: 'flex',
  },
  titleLogo: {
    display: 'inline-flex',
    marginRight: 'auto',
    marginLeft: '0.5rem',
    alignItems: 'center',
    justifyContent: 'start',
    color: 'black',
    fontSize: '1.5rem',
    textTransform: 'uppercase',
    textDecoration: 'none',
    transition: 'opacity 0.15s',
    '& h6': {
      letterSpacing: '4px',
    },
    '&:hover': {
      opacity: '0.75',
    },
  },
  grow: {
    width: '100%',
  },
  sectionDesktop: {
    display: 'none',
    [theme.breakpoints.up('sm')]: {
      display: 'flex',
    },
  },
  sectionMobile: {
    display: 'flex',
    color: 'black',
    [theme.breakpoints.up('sm')]: {
      display: 'none',
    },
  },
  root: {
    flexGrow: 1,
  },
  title: {
    flexGrow: 1,
    display: 'none',
    [theme.breakpoints.up('sm')]: {
      display: 'block',
    },
  },
  search: {
    position: 'relative',
    borderRadius: theme.shape.borderRadius,
    backgroundColor: fade(theme.palette.common.white, 0.15),
    '&:hover': {
      backgroundColor: fade(theme.palette.common.white, 0.25),
    },
    marginLeft: 0,
    marginTop: '6px',
    width: '100%',
    [theme.breakpoints.up('sm')]: {
      marginLeft: theme.spacing(1),
      width: 'auto',
    },
    height: '70%',
  },
  search_mobile: {
    position: 'relative',
    borderRadius: theme.shape.borderRadius,
    backgroundColor: fade(theme.palette.common.white, 0.15),
    '&:hover': {
      backgroundColor: fade(theme.palette.common.white, 0.25),
    },
    marginLeft: 0,
    marginTop: '0px',
    width: '100%',
    [theme.breakpoints.up('sm')]: {
      marginLeft: theme.spacing(1),
      width: 'auto',
    },
    height: '70%',
  },
  circle: {
    height: '25px',
    width: '25px',
  },
}));

const Header: React.FC = () => {
  const { countryId } = useParams<URLParamTypes>();

  const classes = useStyles();
  const dispatch = useDispatch();
  const { t } = useTranslation();

  const language = useSelector((state: IAppState) => state.lang);
  const isLoggedIn = useSelector((state: IAppState) => state.loggedIn);
  const userName = useSelector((state: IAppState) => state.userName);
  const userImage = useSelector((state: IAppState) => state.userImage);

  const avatarSymbol = userName ? userName[0] : 'A';
  const userAvatar = (
    userImage
      ? <Avatar className={classes.circle} src={userImage} />
      : <Avatar className={classes.circle}>{avatarSymbol}</Avatar>
  );

  const avatarElement = !isLoggedIn
    ? <Avatar className={classes.circle} src="/assets/avatar.ico" />
    : userAvatar;

  const [showSignInForm, setShowSignInForm] = React.useState(false);
  const handleCloseSignInForm = () => setShowSignInForm(false);
  const handleShowSignInForm = () => setShowSignInForm(true);

  const [showSignUpForm, setShowSignUpForm] = React.useState(false);
  const handleCloseSignUpForm = () => setShowSignUpForm(false);
  const handleShowSignUpForm = () => setShowSignUpForm(true);

  const [mobileMoreAnchorEl, setMobileMoreAnchorEl] = React.useState<Element>();
  const isMobileMenuOpen = Boolean(mobileMoreAnchorEl);
  const handleMobileMenuOpen = (event: React.SyntheticEvent) => {
    setMobileMoreAnchorEl(event.currentTarget);
  };
  const handleMobileMenuClose = () => {
    setMobileMoreAnchorEl(undefined);
  };

  const [anchorLang, setAnchorLang] = React.useState<Element>();
  const isLangMenuOpen = Boolean(anchorLang);
  const handleLangMenuOpen = (event: React.SyntheticEvent) => {
    setAnchorLang(event.currentTarget);
  };
  const handleLangMenuClose = () => {
    setAnchorLang(undefined);
    handleMobileMenuClose();
  };

  const [anchorAuth, setAnchorAuth] = React.useState<Element>();
  const isAuthMenuOpen = Boolean(anchorAuth);
  const handleAuthMenuOpen = (event: React.SyntheticEvent) => {
    setAnchorAuth(event.currentTarget);
  };
  const handleAuthMenuClose = () => {
    setAnchorAuth(undefined);
    handleMobileMenuClose();
  };

  const [anchorLogOut, setAnchorLogOut] = React.useState<Element>();
  const isLogOutMenuOpen = Boolean(anchorLogOut);
  const handleLogOutMenuOpen = (event: React.SyntheticEvent) => {
    setAnchorLogOut(event.currentTarget);
  };
  const handleLogOutMenuClose = () => {
    setAnchorLogOut(undefined);
    handleMobileMenuClose();
  };

  const handleAuthUserMenuOpen = !isLoggedIn ? handleAuthMenuOpen : handleLogOutMenuOpen;
  const authId = !isLoggedIn ? 'menu-auth' : 'menu-logOut';
  const langId = 'menu-lang';
  const mobileMenuId = 'menu-mobile';

  const logOutMenu = (
    <Menu
      anchorEl={anchorLogOut}
      anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
      id="menu-logOut"
      keepMounted
      transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      open={isLogOutMenuOpen}
      onClose={handleLogOutMenuClose}
    >
      <MenuItem onClick={() => {
        dispatch({ type: 'LOGGED_IN', payload: { loggedIn: false } });
        handleLogOutMenuClose();
      }}
      >
        {t('log_out')}
      </MenuItem>
    </Menu>
  );

  const authMenu = (
    <Menu
      anchorEl={anchorAuth}
      anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
      id="menu-auth"
      keepMounted
      transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      open={isAuthMenuOpen}
      onClose={handleAuthMenuClose}
    >
      <MenuItem
        onClick={() => {
          handleShowSignInForm();
          handleAuthMenuClose();
        }}
      >
        {t('sign_in')}
      </MenuItem>
      <MenuItem
        onClick={() => {
          handleShowSignUpForm();
          handleAuthMenuClose();
        }}
      >
        {t('sign_up')}
      </MenuItem>
    </Menu>
  );

  const langMenu = (
    <Menu
      anchorEl={anchorLang}
      anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
      id={langId}
      keepMounted
      transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      open={isLangMenuOpen}
      onClose={handleLangMenuClose}
    >
      <MenuItem
        onClick={() => {
          dispatch(handleLangChange('en'));
          handleLangMenuClose();
        }}
      >
        {t('english')}
      </MenuItem>
      <MenuItem
        onClick={() => {
          dispatch(handleLangChange('ru'));
          handleLangMenuClose();
        }}
      >
        {t('russian')}
      </MenuItem>
      <MenuItem
        onClick={() => {
          dispatch(handleLangChange('de'));
          handleLangMenuClose();
        }}
      >
        {t('german')}
      </MenuItem>
    </Menu>
  );

  const mobileMenu = (
    <Menu
      anchorEl={mobileMoreAnchorEl}
      anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
      id={mobileMenuId}
      keepMounted
      transformOrigin={{ vertical: 'top', horizontal: 'right' }}
      open={isMobileMenuOpen}
      onClose={handleMobileMenuClose}
    >
      {!countryId && (
      <MenuItem>
        <div className={`${classes.search_mobile}`}>
          <Search />
        </div>
      </MenuItem>
      )}
      <MenuItem onClick={handleLangMenuOpen}>
        <IconButton
          aria-label="Choose lang"
          aria-controls={langId}
          aria-haspopup="true"
          color="inherit"
        >
          <Avatar className={classes.circle} src={`/assets/${language}.ico`} />
        </IconButton>
        <p>{t('language')}</p>
      </MenuItem>
      <MenuItem onClick={handleAuthUserMenuOpen}>
        <IconButton
          aria-label="account of current user"
          aria-controls={authId}
          aria-haspopup="true"
          color="inherit"
        >
          {avatarElement}
        </IconButton>
        <p>{isLoggedIn ? t('log_out') : t('auth')}</p>
      </MenuItem>
    </Menu>
  );

  return (
    <div className={classes.grow}>
      <SignInForm
        isOpen={showSignInForm}
        handleClose={handleCloseSignInForm}
        handleShowSignUpForm={handleShowSignUpForm}
      />
      <SignUpForm
        isOpen={showSignUpForm}
        handleClose={handleCloseSignUpForm}
        handleSnowSignInForm={handleShowSignInForm}
      />
      <AppBar className={classes.appBar} position="static">
        <Toolbar>
          <a href="/" className={classes.titleLogo}>
            <Avatar src="/assets/favicon.svg" />
            <Typography
              variant="h6"
            >
              &nbsp;Travel app
            </Typography>
          </a>
          <div className={classes.sectionDesktop}>
            {!countryId && <Search />}
            <IconButton
              edge="end"
              aria-label="Choose language"
              aria-controls={langId}
              aria-haspopup="true"
              onClick={handleLangMenuOpen}
              color="inherit"
            >
              <Avatar className={classes.circle} src={`/assets/${language}.ico`} />
            </IconButton>
            <IconButton
              edge="end"
              aria-label="account of current user"
              aria-controls={authId}
              aria-haspopup="true"
              onClick={handleAuthUserMenuOpen}
              color="inherit"
            >
              {avatarElement}
            </IconButton>
          </div>
          <div className={classes.sectionMobile}>
            <IconButton
              aria-label="show more"
              aria-controls={mobileMenuId}
              aria-haspopup="true"
              onClick={handleMobileMenuOpen}
              color="inherit"
            >
              <MoreIcon />
            </IconButton>
          </div>
        </Toolbar>
      </AppBar>
      {mobileMenu}
      {authMenu}
      {langMenu}
      {logOutMenu}
    </div>
  );
};

export default Header;
