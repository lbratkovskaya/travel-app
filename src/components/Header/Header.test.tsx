import React from 'react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { shallow, ShallowWrapper } from 'enzyme';
import Header from '../Header';

const middlewares: any = [];
const mockStore = configureStore(middlewares);

describe('Header functional component testing', () => {
  let wrapper: ShallowWrapper<typeof Header>;

  const initialState = {};
  const store = mockStore(initialState);

  beforeEach(() => {
    wrapper = shallow<typeof Header>(<Provider store={store}><Header /></Provider>);
  });

  it('Header must have a correct snapshot', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
