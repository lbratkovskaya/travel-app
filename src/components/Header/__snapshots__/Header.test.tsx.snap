// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header functional component testing Header must have a correct snapshot 1`] = `
"<ContextProvider value={{...}}>
  <Header />
</ContextProvider>"
`;
